### PR TITLE
Optimize empty case in Vec::retain

### PR DESCRIPTION
**Pull request by jcsp**
While profiling some code that happens to call Vec::retain() in a tight loop, I noticed more runtime than expected in retain, even in a bench case where the vector was always empty.  When I wrapped my call to retain in `if !myvec.is_empty()` I saw faster execution compared with doing retain on an empty vector.

On closer inspection, Vec::retain is doing set_len(0) on itself even when the vector is empty, and then resetting the length again in BackshiftOnDrop::drop.

Unscientific screengrab of a flamegraph illustrating how we end up spending time in set_len and drop:
image

### CODE REVIEW

**Comment by conradludgate** [comment]
Slightly more scientific research shows that llvm can produce _less_ instructions with this is_empty check included.

https://godbolt.org/z/jz9Eoqfbc

From looking at it, it is already compiling to a:
```
        mov     r14, qword ptr [rdi + 16]
        test    r14, r14
        je      .LBB1_1
```

which tests if the vec len is 0. Except this `LBB1_1` label does not immediately return and does some other routine instead before finally returning. This label is not reached from any other path in the assembly which looks like a missed optimisation.

File: library/alloc/src/vec/mod.rs [hunk 1]
```diff
@@ -1711,6 +1711,11 @@ impl<T, A: Allocator> Vec<T, A> {
         F: FnMut(&mut T) -> bool,
     {
         let original_len = self.len();
+
+        if original_len == 0 {
+            return;
+        }
```
Comment (tgross35) [inline library/alloc/src/vec/mod.rs:1717]: Could you add a comment saying this allows for better optimization? Just so it doesn't get inadvertently deleted.

**Comment by tgross35** [comment]
Good find! Slightly better view of the asm changes https://godbolt.org/z/r3YvP4s5e, seems to shave about 15% of the assembly on x86, 6% on aarch64.

One request above then LGTM. (please squash back into one commit after you add a comment)

File: library/alloc/src/vec/mod.rs (hunk 1 above)
Comment (jcsp) [reply library/alloc/src/vec/mod.rs]: Sure, added

**Comment by tgross35** [comment]
Thanks!

r? @tgross35
@bors r+

**Comment by klensy** [comment]
rollup=never ?

**Pull request merged**